gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientWriteChannel.java

```diff
@@ -0,0 +1,199 @@
+    WriteChannel writeChannel =
+        storage.writer(
+            getBlobInfo(resourceId, createOptions),
+            generateWriteOptions(createOptions, channelOptions));
+    writeChannel.setChunkSize(channelOptions.getUploadChunkSize());
+
+    return writeChannel;
+  }
+
+  private class UploadOperation implements Callable<Boolean> {
+
+    // Read end of the pipe.
+    private final InputStream pipeSource;
+    private final StorageResourceId resourceId;
+    private final int MAX_BYTES_PER_MESSAGE = MAX_WRITE_CHUNK_BYTES.getNumber();
+
+    UploadOperation(InputStream pipeSource, StorageResourceId resourceId) {
+      this.resourceId = resourceId;
+      this.pipeSource = pipeSource;
+    }
+
+    @Override
+    public Boolean call() throws Exception {
+      // Try-with-resource will close this end of the pipe so that
+      // the writer at the other end will not hang indefinitely.
+      try (pipeSource) {
+        boolean lastChunk = false;
+        ByteBuffer byteBuffer = ByteBuffer.allocate(MAX_BYTES_PER_MESSAGE);
+        while (!lastChunk) {
+          int remainingCapacity = byteBuffer.remaining();
+          ByteString data =
+              ByteString.readFrom(
+                  ByteStreams.limit(pipeSource, remainingCapacity), remainingCapacity);
+          if (data.size() < remainingCapacity) {
+            lastChunk = true;
+          }
+          byteBuffer.put(data.toByteArray());
+          // switch to read mode
+          byteBuffer.flip();
+          // this could result into partial write
+          writeInternal(byteBuffer);
+          if (!lastChunk) {
+            // compact buffer for further writing
+            byteBuffer.compact();
+          }
+        }
+        // last chunk could be partially written
+        // uploading all bytes of last chunk
+        if (lastChunk && byteBuffer.hasRemaining()) {
+          while (byteBuffer.hasRemaining()) {
+            writeInternal(byteBuffer);
+          }
+        }
+        return true;
+      } catch (Exception e) {
+        throw new IOException(
+            String.format("Error occurred while uploading resource %s", resourceId), e);
+      }
+    }
+  }
+
+  private static BlobWriteOption[] generateWriteOptions(
+      CreateObjectOptions createOptions, AsyncWriteChannelOptions channelOptions) {
+    List<BlobWriteOption> writeOptions = new ArrayList<>();
+
+    writeOptions.add(BlobWriteOption.disableGzipContent());
+    writeOptions.add(BlobWriteOption.generationMatch());
+    if (createOptions.getKmsKeyName() != null) {
+      writeOptions.add(BlobWriteOption.kmsKeyName(createOptions.getKmsKeyName()));
+    }
+    if (channelOptions.isGrpcChecksumsEnabled()) {
+      writeOptions.add(BlobWriteOption.crc32cMatch());
+    }
+    return writeOptions.toArray(new BlobWriteOption[writeOptions.size()]);
+  }
+
+  @Override
+  public void close() throws IOException {
+    try {
+      super.close();
+      writeChannel.close();
```

Comment:
usually cleanup is done in the reverse order of creation. So, I think it is safer to close writeChannel before super.close

Reply:
No it's not. `super.close` is a waited call and ends either all of the data in buffer is written over to gcs (using `writeChannel`). `writeChannel` can and will only be closed when there is nothing to be written.

We can't close `writeChannel` unless we are sure that all of the data in pipe is written over.